File: GoLang/quotes/model/linearService.go
package main

import (
	"context"
	"time"

	pb "github.com/RYANCOAL9999/AI_Go_Proto_File"
	"github.com/gin-gonic/gin"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type LinearGRPCClient struct {
	client pb.LinearServiceClient
}

func NewLinearGRPCClient(addr string) (*LinearGRPCClient, error) {
	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}
	return &LinearGRPCClient{
		client: pb.NewLinearServiceClient(conn),
	}, nil
}

func LinearRegressionResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "linear regression"

		var data pb.LinearRegressionRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.LinearRegressionEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}
}

func RidgeResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "ridge"

		var data pb.LinearRidgeRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.LinearRidgeEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)
	}

}

func RidgeCVResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "ridge cv"

		var data pb.LinearRidgeCVRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.LinearRidgeCVEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}

}

func LassoResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "lasso"

		var data pb.LassoExpressionRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.LassoExpressionEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)
	}

}

func LassoLarsResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "lasso lars"

		var data pb.LassoLarsLassoExpressionRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.LassoLarsLassoExpressionEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}

}

func BayesianRidgeResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "bayesian ridge"

		var data pb.BayesianRidgeRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.BayesianRidgeEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}

}

func TweedieRegressorResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "tweedie regressor"

		var data pb.TweedieRegressorRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.TweedieRegressorEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}

}

func SGDClassifierResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "sgd classifier"

		var data pb.SGDClassifierRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.SGDClassifierEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}

}

func ElasticNetResponse(linearClient *LinearGRPCClient) gin.HandlerFunc {
	return func(c *gin.Context) {

		const event_name = "elastic net"

		var data pb.ElasticNetRequest
		if err := ReadRequestData(c, &data); err != nil {
			HandleBadRequestError(c, err, event_name)
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()

		response, err := linearClient.client.ElasticNetEvent(ctx, &data)

		if err != nil {
			HandleGRPCError(c, err, event_name)
			return
		}

		SendResponse(c, response, event_name)

	}

}

File: Javascript/quotes/model/linearService.js
const PROTO_PATH = __dirname + '../lib/AIProto/linear_expression.proto';
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const { defination } = require('./packageDefinitionpackageDefinition')
const { sendStubRequest, handleResponse, handleGRPCError, handleBadRequestError } = require('./helper')

const packageDefinition = protoLoader.loadSync(
    PROTO_PATH,
    defination
);

const linear_proto = grpc.loadPackageDefinition(packageDefinition).linear_expression;

export const createLinearStub = (addr) => {
    return new linear_proto.LinearService(
        addr, 
        grpc.credentials.createInsecure()
    );
};

export const linearRegressionResponse = (req, res, stub) => {

    const event_name = "linear regression"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `LinearRegressionEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const ridgeResponse = (req, res, stub) => {

    const event_name = "ridge"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `LinearRidgeEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const ridgeCVResponse = (req, res, stub) => {

    const event_name = "ridge cv"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `LinearRidgeCVEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const lassoResponse = (req, res, stub) => {

    const event_name = "lasso"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `LassoExpressionEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const lassoLarsResponse = (req, res, stub) => {

    const event_name = "lasso lars"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `LassoLarsLassoExpressionEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const bayesianRidgeResponse = (req, res, stub) => {

    const event_name = "bayesian ridge"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `BayesianRidgeEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const tweedieRegressorResponse = (req, res, stub) => {

    const event_name = "tweedie regressor"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `TweedieRegressorEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const sgdClassifierResponse = (req, res, stub) => {

    const event_name = "sgd classifier"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `SGDClassifierEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

export const elasticNetResponse = (req, res, stub) => {

    const event_name = "elastic net"

    const request_message = req.body; 

    if(request_message){
        sendStubRequest(stub, `ElasticNetEvent`, request_message)
        .then((response) => handleResponse(response, res))
        .catch((err) => handleGRPCError(err, res, event_name));
    }
    else{
        handleBadRequestError(res, event_name);
    }
};

